engine/rpcserver.go

```diff
@@ -2238,3 +2319,294 @@ func (s *RPCServer) WebsocketSetURL(_ context.Context, r *gctrpc.WebsocketSetURL
+	}
+	for i := range trades {
+		resp.Trades = append(resp.Trades, &gctrpc.SavedTrades{
+			Price:     trades[i].Price,
+			Amount:    trades[i].Amount,
+			Side:      trades[i].Side,
+			Timestamp: trades[i].Timestamp.Unix(),
+			TradeId:   trades[i].TID,
+		})
+	}
+	if len(resp.Trades) == 0 {
+		return nil, fmt.Errorf("request for %v %v trade data between %v and %v and returned no results", r.Exchange, r.AssetType, r.Start, r.End)
+	}
+	return resp, nil
+}
+
+// ConvertTradesToCandles converts trades to candles using the interval requested
+// returns the data too for extra fun scrutiny
+func (s *RPCServer) ConvertTradesToCandles(_ context.Context, r *gctrpc.ConvertTradesToCandlesRequest) (*gctrpc.GetHistoricCandlesResponse, error) {
+	if r.End <= 0 || r.Start <= 0 || r.Exchange == "" || r.Pair == nil || r.AssetType == "" || r.Pair.String() == "" || r.TimeInterval == 0 {
+		return nil, errors.New("invalid arguments received")
+	}
+	exch := GetExchangeByName(r.Exchange)
+	startTime := time.Unix(r.Start, 0)
+	endTime := time.Unix(r.End, 0)
+
+	if exch == nil {
+		return nil, errExchangeNotLoaded
+	}
+	sqlTrades, err := tradesql.GetInRange(r.Exchange, r.AssetType, r.Pair.Base, r.Pair.Quote, startTime, endTime)
+	if err != nil {
+		return nil, err
+	}
+	var trades []trade.Data
+	trades, err = trade.SQLDataToTrade(sqlTrades...)
+	if err != nil {
+		return nil, err
+	}
+	if len(trades) == 0 {
+		return nil, fmt.Errorf("no trades returned from supplied params")
+	}
+	interval := kline.Interval(r.TimeInterval)
+	var klineItem kline.Item
+	klineItem, err = trade.ConvertTradesToCandles(interval, trades...)
+	if err != nil {
+		return nil, err
+	}
+	if len(klineItem.Candles) == 0 {
+		return nil, fmt.Errorf("no candles generated from trades")
+	}
+
+	resp := &gctrpc.GetHistoricCandlesResponse{
+		Exchange: r.Exchange,
+		Pair:     r.Pair,
+		Start:    r.Start,
+		End:      r.End,
+		Interval: interval.String(),
+	}
+	for i := range klineItem.Candles {
+		resp.Candle = append(resp.Candle, &gctrpc.Candle{
+			Time:   klineItem.Candles[i].Time.Unix(),
+			Low:    klineItem.Candles[i].Low,
+			High:   klineItem.Candles[i].High,
+			Open:   klineItem.Candles[i].Open,
+			Close:  klineItem.Candles[i].Close,
+			Volume: klineItem.Candles[i].Volume,
+		})
+	}
+
+	if r.Sync {
+		_, err = kline.StoreInDatabase(&klineItem, r.Force)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return resp, nil
+}
+
+// FindMissingSavedCandleIntervals is used to help determine what candle data is missing
+func (s *RPCServer) FindMissingSavedCandleIntervals(_ context.Context, r *gctrpc.FindMissingCandlePeriodsRequest) (*gctrpc.FindMissingIntervalsResponse, error) {
+	if r.End <= 0 || r.Start <= 0 || r.ExchangeName == "" || r.Pair == nil || r.AssetType == "" || r.Pair.String() == "" || r.Interval <= 0 {
+		return nil, errors.New("invalid arguments received")
+	}
+	startTime := time.Unix(r.Start, 0)
+	endTime := time.Unix(r.End, 0)
+	klineItem, err := kline.LoadFromDatabase(
```

Comment:
Not sure if its worth adding an exchange enabled check on this and FindMissingSavedTradeIntervals below to only allow its usage on exchanges currently enabled by bot